trajoptlib/src/mecanum_trajectory_generator.cpp

```diff
@@ -0,0 +1,449 @@
+        auto& dt = dts.at(index);
+        problem.subject_to(dt >= 0);
+        problem.subject_to(dt <= 3);
+        dt.set_value(sgmt_time / N_sgmt);
+      }
+    }
+  }
+  slp::Variable<double> total_time = std::accumulate(dts.begin(), dts.end(), slp::Variable<double>{0.0});
+
+  // Penalty: ∫(v²ω²)dt
+  const double separation_penalty_weight = 0.0;
+  slp::Variable<double> separation_penalty = 0.0;
+
+  for (size_t index = 0; index < samp_tot; ++index) {
+    auto v_trans_sq = vx.at(index) * vx.at(index) + vy.at(index) * vy.at(index);
+    auto omega_sq = ω.at(index) * ω.at(index);
+    separation_penalty += v_trans_sq * omega_sq * dts.at(index);
+  }
+
+  // Penalty: ∫ω²dt
+  const double rotation_penalty_weight = 0.0; // todo
+  slp::Variable<double> rotation_penalty = 0.0;
+  for (size_t index = 0; index < samp_tot; ++index) {
+    rotation_penalty += ω.at(index) * ω.at(index) * dts.at(index);
+  }
+
+  slp::Variable<double> cost = total_time;
+  if (separation_penalty_weight > 0.0) {
+    cost += separation_penalty_weight * separation_penalty;
+  }
+  if (rotation_penalty_weight > 0.0) {
+    cost += rotation_penalty_weight * rotation_penalty;
+  }
+  problem.minimize(cost);
+
+  // apply kinematics constraints
+  for (size_t wpt_index = 0; wpt_index < wpt_cnt - 1; ++wpt_index) {
+    size_t N_sgmt = Ns.at(wpt_index);
+
+    for (size_t sample_index = 0; sample_index < N_sgmt; ++sample_index) {
+      size_t index = get_index(Ns, wpt_index, sample_index);
+
+      Translation2v<double> x_k{x.at(index), y.at(index)};
+      Translation2v<double> x_k_1{x.at(index + 1), y.at(index + 1)};
+
+      Rotation2v<double> θ_k{cosθ.at(index), sinθ.at(index)};
+      Rotation2v<double> θ_k_1{cosθ.at(index + 1), sinθ.at(index + 1)};
+
+      Translation2v<double> v_k{vx.at(index), vy.at(index)};
+      Translation2v<double> v_k_1{vx.at(index + 1), vy.at(index + 1)};
+
+      auto ω_k = ω.at(index);
+      auto ω_k_1 = ω.at(index + 1);
+
+      Translation2v<double> a_k{ax.at(index), ay.at(index)};
+      Translation2v<double> a_k_1{ax.at(index + 1), ay.at(index + 1)};
+
+      auto α_k = α.at(index);
+      auto α_k_1 = α.at(index + 1);
+
+      auto dt_k = dts.at(index);
+      if (sample_index < N_sgmt - 1) {
+        auto dt_k_1 = dts.at(index + 1);
+        problem.subject_to(dt_k_1 == dt_k);
+      }
+
+      // xₖ₊₁ = xₖ + vₖt + 1/2aₖt²
+      // θₖ₊₁ = θₖ + ωₖt
+      // vₖ₊₁ = vₖ + aₖt
+      // ωₖ₊₁ = ωₖ + αₖt
+      problem.subject_to(x_k_1 == x_k + v_k * dt_k + a_k * 0.5 * dt_k * dt_k);
+      problem.subject_to((θ_k_1 - θ_k) == Rotation2v<double>{ω_k * dt_k});
+      problem.subject_to(v_k_1 == v_k + a_k * dt_k);
+      problem.subject_to(ω_k_1 == ω_k + α_k * dt_k);
+    }
+  }
+
+  const double s = 1.0 / std::sqrt(2.0);
+
+  const double strafe_eff = path.drivetrain.strafe_efficiency;
+  std::vector<double> fx_coeff = {s, s, s, s};
+  std::vector<double> fy_coeff = {-s * strafe_eff, -s * strafe_eff, s * strafe_eff, s * strafe_eff};
```

Comment:
It's better to stack-allocate arrays of known size. You'll need to add `#include <array>` at the top of the file for this.
```suggestion
  std::array fx_coeff{s, s, s, s};
  std::array fy_coeff{-s * strafe_eff, -s * strafe_eff, s * strafe_eff, s * strafe_eff};
```